File: PnL/src/pandas/merge Variations tracking.py
import pandas as pd
from pathlib import Path
from janitor import clean_names


# Path
path = Path(__file__).parent.parent.parent


# Functions
def read_multiple_files(list_of_files):
    dataframes = [
        pd.read_excel(
            file,
            sheet_name="Seasonalized Variations",
            skiprows=15,
            nrows=240,
            usecols="A:V",
            # thousands=",",
            # dtype="str",
            dtype={
                "Unnamed: 0": str,  # Plant
                "Unnamed: 1": str,  # Outlet
            },
        ).assign(source=file)
        for file in list_of_files
    ]

    # Add a new column with filename to each DataFrame
    for i, df in enumerate(dataframes):
        df["source"] = list_of_files[i].stem

    # Merge the list of DataFrames into a single DataFrame
    df = pd.concat(dataframes)

    # reorder columns
    df = df[["source"] + [col for col in df.columns if col not in ["source"]]]
    return df


def read_poc(filename):
    df = pd.read_csv(filename, dtype="str")  # .clean_names()
    return df


def read_category_item(filename):
    df = pd.read_csv(filename, dtype="str")  # .clean_names()
    return df


def process_columns(df):
    # clean column names
    # Remove "\n" in the column names
    df.columns = df.columns.str.replace("\n", " ")
    # remove columns
    df = df.drop(columns=["Unnamed: 2", "Unnamed: 3", "Unnamed: 4"])
    # rename columns
    df = df.rename(
        columns={
            "Unnamed: 0": "Plant",
            "Unnamed: 1": "Outlet",
            "Unnamed: 5": "Items",
        }
    )
    # df = df.clean_names()

    return df


def filter_rows(df):
    df = df[df["Plant"] == "242"]
    return df


def process_text(df):
    df["source"] = df["source"].str.extract(r"(CW\d{2})")
    return df


def join_with_poc(df, df_poc):
    df_merged = pd.merge(df, df_poc, on=["Plant", "Outlet"], how="left")
    return df_merged


def join_with_category(df, df_category):
    df_merged = pd.merge(df, df_category, on="Items", how="left")
    return df_merged


def main():

    # Path
    data_path = path / "data" / "Variations tracking"

    # Filenames
    meta_poc = path / "meta" / "POC.csv"
    meta_category = path / "meta" / "Variations items.csv"
    output_file = path / "output" / "Variations trackings.csv"

    # Input data: List of multiple text files
    xls_files = [
        file
        for file in data_path.iterdir()
        if file.is_file() and file.suffix == ".xlsx"
    ]

    # Read data
    df = read_multiple_files(xls_files)
    df_poc = read_poc(meta_poc)
    df_category = read_category_item(meta_category)

    # Process data
    df = (
        df.pipe(process_columns)
        .pipe(filter_rows)
        .pipe(process_text)
        .pipe(join_with_poc, df_poc)
        .pipe(join_with_category, df_category)
    )

    # Write data
    df.to_csv(output_file, index=False)
    print("A file is created")


if __name__ == "__main__":
    main()

File: PnL/src/dplyr/merge Variations tracking.R
library(dplyr)
library(readr)
library(purrr)
library(stringr)
library(readxl)
library(fs)
library(janitor)
library(here)


# Path
path <- here("PnL")


# Functions
read_multiple_files <- function(list_of_files) {
  df <- list_of_files |>
    map_dfr(read_excel,
      sheet = "Seasonalized Variations",
      range = "A16:Y252",
      .id = "source"
    )
  return(df)
}

process_columns <- function(df) {
  # Select columns and remove NA rows
  df <- df |>
    select(-c(4:6, 24:26)) |>
    rename(c(
      Plant  = "...1",
      Outlet = "...2",
      Items  = "...6"
    ))
  return(df)
}

filter_rows <- function(df) {
  df <- df |>
    filter(!is.na(.data[["plant"]]))
  return(df)
}

process_text <- function(df, path_name) {
  # Remove unnecessary text
  df <- df |>
    mutate(
      source = str_remove(.data[["source"]], "_2023"),
      source = str_remove(.data[["source"]], ".xlsx"),
      source = str_remove(.data[["source"]], path_name)
    )
  return(df)
}

poc <- function(filename) {
  df <- read_csv(filename, show_col_types = FALSE) |> clean_names()
  return(df)
}

join_with_poc <- function(df, filename) {
  # Add POC and Variations item information
  df <- df |>
    left_join(poc(filename), by = c("plant", "outlet"))
  return(df)
}

variations_item <- function(filename) {
  df <- read_csv(filename, show_col_types = FALSE) |> clean_names()
  return(df)
}

join_with_variations <- function(df, filename) {
  df <- df |>
    left_join(variations_item(filename), by = "items")
  return(df)
}

main <- function() {
  # Variables
  path_name <- paste0(
    path, "/data/Variations tracking/variations_tracking_ICH_"
  )

  # Path
  data_path <- here(path, "data", "Variations tracking")

  # Filenames
  xls_files <- dir_ls(data_path, regexp = "\\.xlsx$")
  meta_poc <- here(path, "meta", "POC.csv")
  meta_file <- here(path, "meta", "Variations items.csv")
  output_file <- here(path, "output", "Variations trackings.csv")

  # Read data
  df <- read_multiple_files(xls_files)

  # Process data
  df <- df |>
    process_columns() |>
    clean_names() |>
    filter_rows() |>
    process_text(path_name) |>
    join_with_poc(meta_poc) |>
    join_with_variations(meta_file)

  # Write data
  write_csv(df, output_file)
  print("A file is created")
}

main()
